core/nightshade/src/nightshade.rs

```diff
@@ -0,0 +1,323 @@
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum NSResult {
+    Success,
+    Finalize(Vec<CryptoHash>),
+    Retrieve(Vec<CryptoHash>),
+    Error(String),
+}
+
+impl NSResult {
+    #[allow(dead_code)]
+    fn is_success(&self) -> bool {
+        match &self {
+            NSResult::Success => true,
+            _ => false
+        }
+    }
+
+    #[allow(dead_code)]
+    fn is_finalize(&self) -> bool {
+        match &self {
+            NSResult::Finalize(_) => true,
+            _ => false
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct Message<P> {
+    pub author: AuthorityId,
+    pub parents: Vec<CryptoHash>,
+    pub data: P,
+}
+
+impl<P: Payload> Message<P> {
+    pub fn calculate_hash(&self) -> CryptoHash {
+        hash(&self.encode().expect("Serialization failed"))
+    }
+}
+
+#[derive(Debug)]
+struct Node<P: Payload> {
+    message: Message<P>,
+    depth: i64,
+    endorses: AuthorityId,
+    last_depth: Vec<i64>,
+    max_score: Vec<u64>,
+    max_confidence: Vec<u64>,
+}
+
+pub struct Nightshade<P: Payload> {
+    owner_id: usize,
+    num_authorities: usize,
+    nodes: HashMap<CryptoHash, Node<P>>,
+    nodes_per_author: Vec<HashMap<i64, CryptoHash>>,
+    global_last_depth: Vec<i64>,
+    tips: HashSet<CryptoHash>,
+    pending_messages: HashMap<CryptoHash, (Message<P>, HashSet<CryptoHash>)>,
+    missing_hashes: HashMap<CryptoHash, HashSet<CryptoHash>>,
+}
+
+impl<P: Payload> Nightshade<P> {
+    pub fn new(owner_id: usize, num_authorities: usize) -> Self {
+        Nightshade {
+            owner_id,
+            num_authorities,
+            nodes: Default::default(),
+            nodes_per_author: vec![Default::default(); num_authorities],
+            global_last_depth: vec![-1; num_authorities],
+            pending_messages: Default::default(),
+            missing_hashes: Default::default(),
+            tips: Default::default(),
+        }
+    }
+
+    fn filter_candidates(&self, candidates: Vec<usize>, scores: &Vec<u64>) -> Vec<usize> {
+        let mut result = vec![];
+        for &candidate in candidates.iter() {
+            if result.is_empty() || scores[candidate] == scores[result[0]] {
+                result.push(candidate);
+            } else if scores[candidate] > scores[result[0]] {
+                result = vec![candidate];
+            }
+        }
+        result
+    }
+
+    fn update_result(&self, current_result: NSResult, new_result: NSResult) -> NSResult {
+        match new_result {
+            NSResult::Finalize(r) => match current_result {
+                NSResult::Finalize(mut v) => { v.extend(r); NSResult::Finalize(v) },
+                _ => { NSResult::Finalize(r) }
+            },
+            other => other
+        }
+    }
+
+    pub fn process_message(&mut self, message: Message<P>) -> NSResult {
+        let hash = message.calculate_hash();
+        if self.nodes.contains_key(&hash) || self.pending_messages.contains_key(&hash) {
```

Comment:
With `Borrow` this will become `self.nodes.contains(&message) || self.pending_messages.contains_key(&message)`
Notice how we can lookup in `nodes` using `message` because their borrow returns the same type.